File: AverageOfLevelsInBinraryTree_637.js
/*
	description:
		Given a non-empty binrary tree, return the average value
		of the nodes on each level in the form of a array.
		
	Example:
		Input:
		    3
		   / \
		  9  20
		    /  \
		   15   7
		
		Output: [3, 14.5, 11]
		
		Explanation:
			The average value of nodes on level 0 is 3,  
			on level 1 is 14.5, and on level 2 is 11. 
			Hence return [3, 14.5, 11].
		
		Note:
			1.The range of node's value is in the range of 32-bit iterger.
*/			


/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 
 /* the thinking of solution is making two queues.
	the first calculate current layer of nodes.
	the second records next layer nodes.
	loop calculates sum of each node via poping from
	first queue;and pushs the children nodes into if
	one node has children nodes.
	
	reference:https://leetcode.com/problems/average-of-levels-in-binary-tree/solution/
*/	
var averageOfLevels = function(root) {
    var number = [];
	
	var queue = [], tmp = [];
	
	queue.push(root);
	while(!queue.empty()){
		var sumOfLayer = 0, countOfLayer = 0;
		while(!queue.empty()){
			var n = queue.pop;
			sumOfLayer += n.val;
			countOfLayer++;
			if(n.left != null)
				tmp.push(n.left);
			if(n.right != null)
				tmp.push(n.right);
		}
		queue = tmp;
		number.push(sumOfLayer/countOfLayer);
	}
	return number;
};

File: AverageOfLevelsInBinraryTree_637.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */


class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Level order a list of lists of integer
     */
public:
    vector<vector<int>> levelOrder(TreeNode *root) {
         vector<vector<int>>result;
        if (root == NULL) {
            return  result;
        }
        // write your code here

        deque<TreeNode *>oneQueue;
        TreeNode *head = root;
        oneQueue.push_back(head);
		
        int currentLevel = 0;
        int currentLevelInternalCount = 1;
        int nextLevelInternalCount = 0;
        int count = 0;
        vector<int>levelNums;
		
		
        while(!oneQueue.empty()) {
            head = oneQueue.front();
            oneQueue.pop_front();
            count++;
           levelNums.push_back(head->val);
            if (head->left != NULL) {
                oneQueue.push_back(head->left);
                nextLevelInternalCount++;
            }
            if(head->right != NULL) {
                oneQueue.push_back(head->right);
                nextLevelInternalCount++;
            }
            if(count == currentLevelInternalCount) {
                vector<int> a = levelNums;
                //vector<int>a(2,0);
                result.push_back(a);
                levelNums.clear();
                count = 0;
                currentLevelInternalCount = nextLevelInternalCount;
                nextLevelInternalCount = 0;
            }
        }
        return result;
    }
};